src/rpc/RPCClient.ts

```diff
@@ -78,22 +86,24 @@ class RPCClient<M extends ClientManifest> {
   >;
   protected callerTypes: Record<string, HandlerType>;
   // Method proxies
+  public readonly defaultTimeout: number;
   public readonly methodsProxy = new Proxy(
     {},
     {
       get: (_, method) => {
         if (typeof method === 'symbol') throw never();
         switch (this.callerTypes[method]) {
           case 'UNARY':
-            return (params) => this.unaryCaller(method, params);
+            return (params, ctx) => this.unaryCaller(method, params, ctx);
           case 'SERVER':
-            return (params) => this.serverStreamCaller(method, params);
+            return (params, ctx) =>
+              this.serverStreamCaller(method, params, ctx);
           case 'CLIENT':
-            return () => this.clientStreamCaller(method);
+            return (ctx) => this.clientStreamCaller(method, ctx);
           case 'DUPLEX':
-            return () => this.duplexStreamCaller(method);
+            return (ctx) => this.duplexStreamCaller(method, ctx);
           case 'RAW':
-            return (header) => this.rawStreamCaller(method, header);
+            return (header, ctx) => this.rawStreamCaller(method, header, ctx);
```

Comment:
I'm refreshing the timeout per message by design. It's too cheap for a bad actor to hold a connection open if it refreshes per chunk.

If I wanted to, even for a unary call I could send a single byte every 50ish seconds to hold a connection open. I wouldn't have to send a valid message at all to do this so long as it's valid JSON for parsing.  I could do this for a very long time until I hit the message parsing buffer limit.

Having it per message means you'd have to at least send valid messages and conform to the protocol of the RPC method you're using.

Reply:
What about raw streams?

Reply:
Raw streams doesn't support the refreshing. It will be up to the handlers to implement the refreshing logic. Mostly for the same reason above.

Reply:
> If I wanted to, even for a unary call I could send a single byte every 50ish seconds to hold a connection open. I wouldn't have to send a valid message at all to do this so long as it's valid JSON for parsing. I could do this for a very long time until I hit the message parsing buffer limit.

This isn't really an issue.

The purpose of timeouts are is not to deal with DOS. It's to deal with idle clients/idle servers.

What you're describing is perfectly fine.

Reply:
Ok, I'll have to make some changes then.

Reply:
Raw streams cannot have timeouts... because the transport layer is already doing this. So this is not relevant anymore. But we need comments inside explaining why.

Reply:
If you are resetting the timer every time you receive a message and every time you respond. In this case resetting before the response is sent.

You have to deal with the possibility that the other side is not consuming any of the data you send.

From the client's perspective the clock starts as soon as you attempt a call. Meaning that if the server is "SLOW" then eventually it will be aborted.

From the server's perspective, if you are responding with data, the client may be SLOW. Then in that case, you want to make sure that the time spent responding is also tracked.

Reply:
We should do some tests that test race conditions here.. just to see how client timer and server timer interacts with large request and large responses. Imagine the request/response messages are the size of 1 MiB.

Message size should be considered here.

What we want to prevent is memory overflow here. Even in the case of raw streams.

Can you make sure the buffer size for JSON parsing is part of the `src/config.ts` as well.

Reply:
Buffer limit for parsing has been added to the config.

As for testing race conditions, I don't see the need right now. The difference between the client timing itself out or the server doing it is pretty minor. If the client times itself out then the stream ends with a timeout error.  If the server does it then the client sees the connection ending early and the streams end with a different error, connection ended early I think.

I could add more informative websocket stream ending with error codes. But I don't see the need right now. I can create issues for both timeout race testing and more informative error codes for websockets ending.

Reply:
> As for testing race conditions, I don't see the need right now. The difference between the client timing itself out or the server doing it is pretty minor. If the client times itself out then the stream ends with a timeout error. If the server does it then the client sees the connection ending early and the streams end with a different error, connection ended early I think.

Can you make sure there are tests covering when the client timesout VS when the server timesout. Please give the name of the test here.

> I could add more informative websocket stream ending with error codes. But I don't see the need right now. I can create issues for both timeout race testing and more informative error codes for websockets ending.

This needs to be designed along with the mapping functions use in js-quic. Can you point out where in our spec that error codes is being specified?

Reply:
RPCServer has two tests, if the stream fails and plenty of timeout tests.
* `'should emit stream error if input stream fails'`
* `'should emit stream error if output stream fails'`

The `RPCClient` doesn't have any tests. But it doesn't do anything complex here. if it's timing out then it cancels the stream, if the stream ends early then the stream throws. Both cases is the stream throwing an error. 

I don't expect any odd interactions happening depending on race conditions with timeouts. So I didn't write any integration tests with that in mind. 

As for the WS error codes. It's not in the spec. There is only 1 code I'm using `4001` as a generic catch all. We don't use the closing frames to convey any errors. the only time it would contain messages is if we ended the connection abruptly and we only use that information for debug messages.

We could expand this feature out but I don't think it's needed.  It's also out of scope for this PR.

Reply:
Can you change to 4000 as the generic error code. 4001 and above should be specific codes. This matches the `0` being used in QUIC.